File: ui/comms/serialreader.py
from PySide.QtCore import QThread, Signal
import struct
import comms
import serial
import time

class SerialReader(QThread):
    is_finished = False
    on_config_received = Signal(object)
    on_gyro_acc_received = Signal(object)
    on_success = Signal()
    on_failure = Signal()

    def __init__(self, serial_port):
        self.serial_port = serial_port
        super(SerialReader, self).__init__()
        self.start()

    def run(self):
        code = 0
        state = 0
        crc = 0
        data_expected_length = 0
        data_received_length = 0
        data_buffer = bytes()
        data = 0

        while self.is_finished == False:
            if comms.SerialManager().port_is_open():
                try:
                    data = self.serial_port.read()
                except serial.serialutil.SerialException:
                    print("error - closing serial port")
                    comms.SerialManager().close_port()
            else:
                time.sleep(0.2)
                continue

            if data == b'':
                # this is for when the serial port times out.
                # does this have a chance of breaking things if the data is ever empty?
                continue

            if (state == 0):
                if (ord(data) == comms.PACKET_HEADER1):
                    state += 1
                    self.log("got header1")
            elif (state == 1):
                if (ord(data) == comms.PACKET_HEADER2):
                    state += 1
                    self.log("got header2")
                else:
                    state = 0
            elif (state == 2):
                code = ord(data)
                self.log("code: {}".format(code))
                crc ^= ord(data)
                state += 1
            elif (state == 3):
                data_expected_length = ord(data)
                crc ^= ord(data)
                state += 1
            elif (state == 4):
                data_expected_length |= (ord(data) << 8)
                crc ^= ord(data)
                state += 1
                self.log("data length: {}".format(data_expected_length))
            elif (state == 5):
                data_buffer += data
                crc ^= ord(data)
                data_received_length += 1
                if (data_received_length >= data_expected_length):
                    state += 1
            elif (state == 6):
                self.log("expected: {}".format(data_expected_length))
                self.log("received: {}".format(data_received_length))

                if (crc == ord(data)):
                    self.log("crc correct!")

                    if (code == comms.REQUEST_CONFIG):
                        self.unpack_config_data(data_buffer)
                    elif (code == comms.REQUEST_GYRO_ACC):
                        self.unpack_gyro_acc_data(data_buffer)
                    elif (code == comms.INFO_SUCCESS):
                        self.on_success.emit()
                    elif (code == comms.INFO_FAILURE):
                        self.on_failure.emit()
                    else:
                        self.log("unrecognized code: {}".format(code), True)
                else:
                    self.log("calc crc: {}".format(crc))
                    self.log("got crc: {}".format(ord(data)))
                    self.log("crc bad!")
                state = 0
                crc = 0
                data_expected_length = 0
                data_received_length = 0
                data_buffer = bytes()

    def unpack_gyro_acc_data(self, data_buffer):
        self.log("unpacking gyro acc data")
        data = struct.unpack("< fff fff fff hhh hhh fff fff", data_buffer)
        self.log("recieved gyro acc data: {}".format(data))
        self.on_gyro_acc_received.emit(data)

    def unpack_config_data(self, data_buffer):
        self.log("recieved config data")
        self.on_config_received.emit(data_buffer)

    def finished(self):
        self.is_finished = True

    def log(self, text, important=False):
        if not important: return
        print("[SerialReader] {}".format(text))

File: js_ui/app/serial/SerialReader.js
import * as serialCodes from './serialCodes'
import { parseStruct } from "./struct";
import * as structLayouts from "./structLayouts";

class SerialReader {
  code = 0;
  state = 0;
  incomingCrc = 0;
  dataExpectedLength = 0;
  dataReceivedLength = 0;
  dataBuffer = null;
  dataBufferView = null;

  constructor(dataReceived) {
    this.dataReceived = dataReceived;
  }

  processCommand(data) {
    switch(this.state) {
      case 0:
        if (data === serialCodes.PACKET_HEADER1) {
          this.state++;
          this.log("got header 1");
        }
        break;
      case 1:
        if (data === serialCodes.PACKET_HEADER2) {
          this.state++;
          this.log("got header 2");
        } else {
          this.state = 0;
        }
        break;
      case 2:
        this.code = data;
        this.incomingCrc = data;
        this.state++;
        this.log(`got code: ${data}`);
        break;
      case 3:  // Data length LSB
        this.dataExpectedLength = data;
        this.incomingCrc ^= data;
        this.state++;
        break;
      case 4:  // Data length MSB
        this.dataExpectedLength |= (data << 8);
        this.incomingCrc ^= data;
        this.log(`data length: ${this.dataExpectedLength}`);
        this.dataBuffer = new ArrayBuffer(this.dataExpectedLength);
        this.dataBufferView = new Uint8Array(this.dataBuffer);
        this.state++;
        break;
      case 5:
        this.dataBufferView[this.dataReceivedLength++] = data;
        this.incomingCrc ^= data;
        if (this.dataReceivedLength >= this.dataExpectedLength) {
          this.state++;
        }
        break;
      case 6:
        this.log(`data length received: ${this.dataReceivedLength}`);
        this.log(`incoming crc: ${data} -- calc'd crc: ${this.incomingCrc}`);

        if (this.incomingCrc === data) {
          this.log("crc match!");

          var response;

          switch(this.code) {
            case serialCodes.REQUEST_CONFIG:
              response = parseStruct(this.dataBuffer, structLayouts.config);
              break;
            case serialCodes.REQUEST_GYRO_ACC:
              response = parseStruct(this.dataBuffer, structLayouts.gyroAcc);
              break;
            case serialCodes.REQUEST_RC:
              response = parseStruct(this.dataBuffer, structLayouts.rcData);
              break;
            case serialCodes.INFO_SUCCESS:
              console.log("controller responded with success!");
              break;
            case serialCodes.INFO_FAILURE:
              console.log("controller responded with failure");
              break;
            default:
              console.log(`unknown code ${this.code}`);
          }

          this.dataReceived(this.code, response);

        } else {
          console.log("bad crc!");
        }

        this.code = 0;
        this.state = 0;
        this.incomingCrc = 0;
        this.dataExpectedLength = 0;
        this.dataReceivedLength = 0;
        this.dataBuffer = null;
        this.dataBufferView = null;
        break;
    }
  }

  log(text) {
    return;
    console.log(text);
  }
}

export default SerialReader;
